### DISCUSSION TITLE
I made some changes to Unit 7. Sharing those changes here for comments or reuse.

### DISCUSSION THREAD

**Discussion by jdonwells** [TEALSK12/2nd-semester-introduction-to-computer-science#390] [Show and tell]
The problem I saw with Unit 7 is the way it starts. We show them how to create empty classes and then populate the instances with attributes individually outside of the class. If you have spent any time in this discussion group you can imagine that doesn't work for me.

7.01 So for the first lesson I went over class creation, attributes, and methods. The lab was changed to just ask them to create a class and a method that sets the attributes. Essentially an init method but not __init__ yet. https://docs.google.com/presentation/d/1v-03EpGBxwvYl3tUei4xTI2iQWWns9wZye3NKDP4htc/edit?usp=sharing

7.02 is not changed. They learn about a better way to initialize with the __init__ method.

7.03 Mostly unchanged. The lab is changed to just ask them to create a __str__ method. https://docs.google.com/presentation/d/1wnYQXib_DGD5hIXH15D7w9Lyq8LRm9-KHa9iMsqq52c/edit?usp=sharing
Notice that we are using a piece of our Zork game that was moved to Unit 6 as the basis for the lab.

7.04 is not changed. They learn about inheritance.

7.05 is a new lesson. In this lesson I put everything they have learned together with super() and double dispatch. They are then given a lab to create __add__ methods for two classes. Basically the Kangaroo lab from 7.03 but using double dispatch. They have items and backpacks. You can add an item to a backpack or a backpack to an item. However, you may not add two items or two backpacks together. That will print an error. https://docs.google.com/presentation/d/1fLtSQPkeQyVOJTsL5sAyODp6V6nlFfR7Nh-sFrFnBeA/edit?usp=sharing

**Comment by Dan-TEALS** [comment]
Hi Don, I appreciate your viewpoint on this. I agree that, while the approach in the curriculum may be "Pythonic," it's not good OOP. The structure of unit 7 reflects the approach taken by Allen Downey in the _Think Python_ text that was the basis for the Python curriculum.

To be honest, with the curriculum going into maintenance mode, I'm not sure we're going to change the current approach, but I will keep your suggestions in mind as I update unit 7.

**Reply by jdonwells** [reply]
I would not call it Pythonic.

I did some research on _Think Python_ it started out as a Java book. There is an old saying. You can write FORTRAN code in any language.

**Reply by Dan-TEALS** [reply]
The forward to _Think Python_ points out that it was a rewrite of the Java textbook _How to Think Like a Computer Scientist_ that Allen Downey wrote after realizing that Java was too confusing for students starting out in programming. For the most part, _Think Python_ reflects the style and content of the Python.org tutorials. In fact, the use of an empty class is shown in the Python.org tutorials as a way to create the equivalent of a C struct.

I do agree that we shouldn't lead with an example of adding attributes to an empty class. It's confusing and poor OOP, but it is definitely something embraced as one approach to using classes by the Python Foundation. I'll definitely be taking a close look at lesson 7.1 as we revise it.

Thanks, as always. for your valuable feedback, Don.
